### PR TITLE
fix(BusinessObjective): RHICOMPL-874 prevent duplications

### PR DESCRIPTION
**Pull request by skateman**
## Secure Coding Practices Checklist GitHub Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Checklist
- [x] Input Validation
- [x] Output Encoding
- [x] Authentication and Password Management
- [x] Session Management
- [x] Access Control
- [x] Cryptographic Practices
- [x] Error Handling and Logging
- [x] Data Protection
- [x] Communication Security
- [x] System Configuration
- [x] Database Security
- [x] File Management
- [x] Memory Management
- [x] General Coding Practices

### CODE REVIEW

File: app/services/duplicate_business_objective_resolver.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+      duplicates = BusinessObjective.select(MAIN_ID_COLUMN, DUPL_ID_COLUMN)
+                                    .group(:title)
+
+      ActiveRecord::Base.connection.execute("
+        UPDATE policies SET business_objective_id = sq.main
+        FROM (#{duplicates.to_sql}) AS sq WHERE business_objective_id = sq.dupl
+      ")
+
+      # This is safe as we resolved the relations with the UPDATE above
+      BusinessObjective.where(id: duplicates.pluck(DUPL_ID_COLUMN)).delete_all
```
Comment (vkrizan) [inline app/services/duplicate_business_objective_resolver.rb:19]: `DUPL_ID_COLUMN` is lacking an order. Delete query might delete BO that the duplicates where set to.

**Review by vkrizan — COMMENTED** [review]
@skateman can you please describe the motivation? The RHICOMPL-827 does not seem to be the correct one.

Since there is a separate table for Business Objectives, for a single text field, it might be a good time to reflect on it. The separation and its own API calls (mainly in GraphQL) had me thinking about having BO per account. That would allow for customers to list BO and add any other metadata to it. Just a thought. If this is not needed, we might think of moving to a flat text field on Policies.

**Comment by skateman** [comment]
Sorry @vkrizan it should be [874](https://issues.redhat.com/browse/RHICOMPL-874)

**Comment by vkrizan** [comment]
@skateman reminder that this is still opened.

File: app/services/duplicate_business_objective_resolver.rb (hunk 1 above)
Comment (skateman) [reply app/services/duplicate_business_objective_resolver.rb:19]: fixed

**Review by vkrizan — APPROVED** [review]
LGTM. 
Be sure to test this on stage with some bigger data set. /cc @Victoremepunto @LightOfHeaven1994

**Comment by InsightsDroid** [comment]
Test results are available in [Ibutsu](https://url.corp.redhat.com/ibutsu-runs). The test run IDs are:
b2427253-9933-4080-bbcc-7e756afa7fcd
bf1f1766-c248-410c-9cd0-6a1c82030d3f

**Pull request merged**